### DISCUSSION TITLE
Relics from C++ in C#: function arguments

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6909] [Language Ideas]
Dear all,

this is another annoying relic of C++. Suppose you have the following function declaration

```C#
void minimize(in double x0, in double y0, in double z0, out double x_min, out double y_min, out double z_min);
```

wouldn't it result in shorter and more readable code if the following syntax was allowed?
```C#
void minimize(in double x0, y0, z0; out double x_min, y_min, z_min);
```

I understand that I could have use a double[3] array from x0s and xmins, but this is not the point here, it is a simple example to demonstrate the syntax issue. How many times we end up with long function declarations due to many arguments ? Pascal and ADA got that right from the very beginning. I would be much happier if syntax like that was supported some time in the future. Do you think it has a chance?

**Comment by CyrusNajmabadi** [comment]
You should not use 'in' in this fashion here. It will only serve to hurt performance.  'in' is for passing large structure as ref-readonly.

**Reply by ghost** [reply]
Thanks for the hint, but had to declare a class before passing it as argument so I used simple types that everyone knowns to illustrate the idea.

**Reply by 333fred** [reply]
Don't pass reference types by `in` either.

**Reply by ghost** [reply]
> Don't pass reference types by `in` either.

I am aware of that 333fred but it is not the point here.

**Comment by CyrusNajmabadi** [comment]
> Do you think it has a chance?

Possibly. But it also doesn't seem to matter much. Bring explicit with the type is simple and readable. If we hit a lot of feedback on this, we might consider it though.

**Reply by ghost** [reply]
Thank you. 

Sometimes, there are so many arguments and the function declarations become so large and there is no way around it because it is an algorithm with specific inputs that mean different things. Anyway, these are just some ideas in case you guys find them interesting, nothing more.

**Reply by CyrusNajmabadi** [reply]
I'd recommend moving away from parameters for that, and use the parameter-object pattern. You could then use a record struct for that purpose.  

This feature does not help with your case. You still have just as many parameters. And you'll likely need to name them at the use site to make sure it's clear and you're not making mistakes. At that point might as well use the better patterns available.

**Comment by theunrepentantgeek** [comment]
> wouldn't it result in shorter and more readable code if the following syntax was allowed?
```
void minimize(in double x0, y0, z0; out double x_min, y_min, z_min)
```

That's shorter, yes. I'd argue that it's more cryptic, not more readable.

C# does allow that kind of shortcut with variables:

``` csharp
double x0, y0, z0;
double x_min, y_min, z_min;
```

But every development team I've worked on has prohibited this style (in any/every language) as prone to error because it's too easy for a future maintainer to make a mistake modifying the method. 

_Aside_: this code seems to be suffering from _primitive obsession_. 

Declare a suitable value type and much of the problem goes away:

``` csharp
void minimize(in Point3D p; out Point3D min)
```

My experience declaring these kinds of value types is that they _always_ simplify code and prevent bugs. The JIT compiler is very good at rendering these into very efficient runtime code, good enough that the gains for the developer are definitely worth it.

**Comment by theunrepentantgeek** [comment]
> How many times we end up with long function declarations due to many arguments ?

Clean code aficionados would suggest the maximum number of arguments to a method is one. 

I think that's extreme - but I would agree that the ideal number is low. If you're finding yourself with a large number (say, anything more than three or four) of method parameters, it's worth reconsidering your approach.

**Reply by bernd5** [reply]
I absolutely agree. Usually I introduce in those cases a `XXXOption` type...

**Reply by ghost** [reply]
> > How many times we end up with long function declarations due to many arguments ?
> 
> Clean code aficionados would suggest the maximum number of arguments to a method is one.
> 
> I think that's extreme - but I would agree that the ideal number is low. If you're finding yourself with a large number (say, anything more than three or four) of method parameters, it's worth reconsidering your approach.

Guys, please, stop repeating general rules written in programming books as general directive or rule of thumb. I am pretty aware of these rules. What you may not be aware, is that there are cases where for clarity and readability you cannot do otherwise. In mathematics you will see a lot of such examples, which you do not find in other areas. There in order to maintain a readable code 1-1 agreement of pseudocode and implementation in some language, you have to maintain the number of variables, notations etc.  Reconsidering the approach in these cases for what? For some heuristic rule informatics people wrote in a book in order to force their students to think twice before unnecessarily pack too many arguments in a function? I do not disagree with the rule. It is a nice thing to keep in mind. But there are cases where it applies and cases where it does not apply. Every rule has its limitations. We should not sacrifice readability for man made rules. 

Algorithms have stopping criteria. Stopping criteria may involve thresholds of updates norms, and max iteration counts of several algorithmic phases. As I said in the beginning of this post, it is not a nice idea in terms of readability to introduce arrays to hold all the integer parameters and all the double parameters in order to reduce arguments in the argument list of the function. Then you have to remember which is threshold is the [0], the [1], the [2], the [3] etc.

**Reply by CyrusNajmabadi** [reply]
> As I said in the beginning of this post, it is not a nice idea in terms of readability to introduce arrays to hold all the integer parameters and all the double parameters in order to reduce arguments in the argument list of the function. 

No one is suggesting using arrays. We have suggested using the parameter object pattern (which you would generally use a record for in c#). 

> Then you have to remember which is threshold is the [0], the [1], the [2], the [3] etc.

No you don't. Because you're not using an array. You're using a record, which uses names for each parameter or member. 

> Guys, please, stop repeating 

@Dhmosthenis a core part of design is considering necessity, as well as evaluating what alternatives to the design exist.  People aren't repeating this because of something they read. They're starting I this as an option because it's a solution we know works given our experience with the language and similar problems.

**Comment by DavidArno** [comment]
```cs
record struct Xyz(double x, double y, double z);

Xyz Minimise(Xyz xyz);
```
Shorter code. No `in`'s; no `out`'s. No new syntax needed. Job done. 😃

**Reply by ghost** [reply]
As I explained, you are not allowed to do that. Do you really think that I would go through all the typing for this post if there was an alternative?

**Reply by DavidArno** [reply]
> As I explained, you are not allowed to do that. 

You are always allowed to do that unless you are in some contrived coding test for an interview 😃. 

As you posted elsewhere, the `in` and `out` here are a distraction and what you want to do is express:
```cs
double Minimize(double x0, double y0, double z0);
```
as
```cs
double Minimize(double x0, y0, z0);
```
instead.

I've had a search back through the discussions and - as far as I can tell - your are the first person to suggest this as an improvement in nearly 7,000 discussions, spanning many years. The reason is simple: people rarely hit the issue you seem concerned with. Typically, if a function has many parameters, they will be of different types, or they will be `params` based if they are multiple parameters of the same type. In situations where there are a fixed number of the same type, they typically will be `XYZ`, `RGB` `HSV` type groups. If they consist of 2-3 parameters, that's not a significant overhead. If one has multiple sets of eg XYZ or larger groups, then grouping them in a record, collection etc then tends to make sense.

Being able to express `double x0, double y0, double z0` as `double x0, y0, z0` doesn't add enough to the language to justify the added complexity it brings both to the language itself and the effort needed to learn it. `record struct Xyz(double x, double y, double z)` is both allowed and sensible. So use these existing features.

**Reply by ghost** [reply]
I do not see any complexity, in fact it is much less complexity and people will be happy that they will have not have to repeat `int,int,int` as many times if a shorter notation does the job. If you do not like it fine. But do not try to convince me it is more complicated. I would be happy to use structures as I told you before, but in this case I cannot. Clearly, there are exceptions to the general rules. In this case, there are a lot of exceptions. As I said, rules you find in books for teaching languages are general guidelines for newbies. They are not always applicable. You have to understand that there are much more important priorities when writing code implementing algorithms for mission critical applications. You see, they realized that in ADA and this is the syntax of ADA long time ago. It is a language for mission critical applications, high reliability systems and projects, such as missiles, reactors and power plants, designed and developed by mature engineers. They realized the importance of this syntax. It is clearly better whether you want to accept it or not. If C# does not want to support it, then I guess we have to live with their decision.

**Reply by HaloFour** [reply]
C# is optimized for the reader, not the writer.  This is because code needs to be reviewed by a human much more frequently than it will be written.  Clarity is the most important feature.  Trying to make the method signatures more terse in extremely specific and arbitrary cases, and against the stylistic guidelines recommended for general users of the language in other contexts is the opposite of making the language more clear.

Furthermore, I find your entire manner of conversation on these forums to be pretty disruptive.  Constantly claiming that alternative approaches cannot be suggested or that guidelines are for "newbs" or that if C# doesn't adopt the stylistic approach of other programming languages that it will remain a toy language does not further your cause.  All it does is make people less likely to want to engage with and consider your proposals.

**Reply by DavidArno** [reply]
> You see, they realized that in ADA and this is the syntax of ADA long time ago. It is a language for mission critical applications, high reliability systems and projects, such as missiles, reactors and power plants, designed and developed by mature engineers

This bit made me chuckle to be honest. I spent a few years writing ADA for various military projects in the early 90's. I'd studied it at college and fell in love with its generics, so was keen to work with it. My first job involved FORTRAN, so ADA seemed a massive leap forward when I got the chance to switch to working with it. A few years later I'd had enough of its verbose syntax, painfully slow compilations and plain daft rules that did little to prevent bugs, switched to other languages and never went near it again.  

Despite well and truly fitting the "mature engineer" label these days, I'll stick with "toy" languages, thanks. 😆

**Reply by ghost** [reply]
I agree with you that some things in ADA are just too much. At the moment I think that the only ADA thing that it is too much is access to types. They could have adopted a much simpler way of define pointers with a `pointer` keyword. Other than that I haven't seen anything that it is too much verbose.

**Reply by CyrusNajmabadi** [reply]
> As I explained, you are not allowed to do that

Yes you are. This is completely legal c#. You are certainly allowed to do this from our perspective. 

You may work with someone who disallows this. But that's not our concern. We're not going to add *new* features because your job doesn't let you use a perfectly good existing feature. Think about the logic there. Say we add the feature, but you (or someone else) comes along and says "but I'm not allowed to use that new feature either. So give me yet another new feature!" 

As a rule, we do not create replacement features for things that already exist just because people voluntarily make the choice to not use them.  That's your own choice.

**Reply by ghost** [reply]
> As a rule, we do not create replacement features for things that already exist just because people voluntarily make the choice to not use them. That's your own choice.

Cyrus, this I understand, but believe me not even you in certain cases would have used a record or a structure as a solution to the problem I mentioned. The only reason I suggested that is because looking the dynamics of C# expansion and adoption of new ideas such as things that are absolutely not necessary in a new Language (not necessary because C# already provides more than many languages) such as (roles, extensions, etc) I suggested a few things which are much easier to fix than inventing a new huge feature (roles, or something similar) to beautify the semantics of the language. That is all. We have discussed it more than it deserves. If you guys opt not to do it, don't. End of discussion. I believe this is the right way to save the developer from unnecessary typing on his keyboard. You do not agree, fine! People will have to live with your choice. They will judge your decision. Introducing a record for things that are totally different in nature and scope of existence to save keyboard strokes and declaration function signatures is not the answer however.

**Reply by CyrusNajmabadi** [reply]
> but believe me not even you in certain cases would have used a record or a structure as a solution to the problem I mentioned.

You'd have to show those cases then.  We can only drive our decisions based on the examples given :)

> Introducing a record for things that are totally different in nature and scope of existence to save keyboard strokes and declaration function signatures is not the answer however.

Why not? :)

It's something people have been doing for decades with very high success.  It's codified into a pattern used across the ecosystem.  If you want to state taht your case is special and shouldn't work the same way then it's incumbent on *you* to make that argument.

**Reply by tannergooding** [reply]
> Introducing a record for things that are totally different in nature and scope of existence to save keyboard strokes and declaration function signatures is not the answer however.

This is also done for performance reasons. At the ABI level (that is the contract defined by the operating system and/or runtime on how parameters are passed between functions) you get around ~4-5 registers for passing values (on modern 64-bit systems; most 32-bit systems always pass all values on the stack). Values that do not fit in register get passed via a reference to shadow copy space on the stack. The same happens when you have more parameters than are allowed to be passed in register and such values gets passed via a shadow copy to the stack. Each parameter is copied to the stack individually and that can impact perf negatively. It is not trivial for such values to be copied in bulk due to layout requirements defined at the ABI level.

Utilizing a record or a struct to pass around related state is therefore a common optimization in many languages. It also allows extending that state without requiring new overloads and allows the various compilers (high level like C# or low level like JIT/AOT) to better optimize and understand your code in turn.

**Reply by ghost** [reply]
@tannergooding,
I am fully aware of that, but what you are not aware of, is that the algorithms and the functions I am talking about may run for 1 hour or more. Passing a few arguments instead of a data structure will not change the overall runtime which will still be 1 hour or more. And as I said to @CyrusNajmabadi, I respect your efforts and I am trying here to suggest a few things that will make our coding faster and cleaner. I am here because I saw some dynamics in your efforts. Otherwise I would not bother. 

I totally understand that these suggestions are not the top in your TODO list. This I totally accept and I support it and I fully and wholeheartedly agree with you. Of course we can live without these suggestions for several decades, that is why I chose the title as `Relics of C++`. To recognize that whether we keep them or not these are relics and nothing more than relics. Of course C# is a C-based language and as such it has to inherit something to honor its name. But should it? With all the efforts you are investing creating a better language, should C# support to the end all the relics of C, C++?

But do not try to convince me please, that `ref` is better in terms of clean semantics than `in out` or that `*` is better than `pointer` because of `T*** var`. Whoever uses `T***` he does not know what he is doing. So `pointer pointer pointer T` is not an argument. You can always invent a better syntax for that such as `pointer^3` or `pointer^2` instead of the cubersome `pointer pointer pointer`. 

I am closing this discussion by repeating once again that I respect and admire your efforts and your drive. I know that you are capable people and you are doing good. But believe me, if you do not see it now, you will see it after 10-20 years of experience: 

_**The details and the attention to details makes the difference.**_

**Reply by CyrusNajmabadi** [reply]
> Passing a few arguments instead of a data structure will not change the overall runtime which will still be 1 hour or more.

It absolutely may, depending on where the hotspots are.  

Regardless, that's not even relevant here.  This is a discussion about ease of use writing and reading these signatures.  The parameter-object pattern is an industry standard pattern that works exceptionally well for this case.

You may not like that.  Or you may have rules that disallow you from using it.  However, discussion of this alternative is a *core part* of any language proposal assessment.  In other words, we have to ask ourselves "why would we add feature Y, when feature X already exists and seems to solve exactly the problem the user is presenting?". 

> I totally understand that these suggestions are not the top in your TODO list. 

This is not the concern.  The concern is whether or not the actual feature request is a *good* request for hte language to take.  Currently, all of your proposals have come with very significant drawbacks from a language perspective.  They're also not done a good job motivating their need, especially in terms of comparing to existing options. At the very least, when people suggest alternatives, you must give sensible reasoning about why those alternatives are not ok.  Simply rejecting them by fiat with false claims (like "you are not allowed to do that") has no weight here as to us that statement is simply not true.

> should C# support to the end all the relics of C, C++?

There is much of those languages we do not support.  And we feel no need to be beholden to them.  That said, you are talking about features we have that we legitimately have no issue with.  Over multiple decades, and feedback sources from thousands to of users, groups and companies, we have not heard or observed problems in this area.  So when you come in insisting that there are serious issues here that need to be addressed, that falls *very flat* for us as it does not align at all with our own experiences.  As such, it makes it feel much more like this is just a single person complaining about things they personally don't like, as opposed to serious issues that impact the ecosystem as a whole that we need to focus on.

> You can always invent a better syntax

But why would we?  We have a syntax that works well *today*.  You don't like it, but that has very little weight for us since the rest of hte ecosystem is fine with it.  Consider if we had to just change every piece of syntax just because a single person came along complaining about.  We'd have 10+ alternatives for every piece of syntax in the language just to appease those critics.  That's not a viable way to direct hte language at all.
